contracts/peripheral/YieldDirector.sol

```diff
@@ -3,355 +3,489 @@ pragma solidity ^0.8.10;
+        } else {
+            _withdrawPrincipal(depositId_, amount_, msg.sender);
+            amountWithdrawn = amount_;
+        }
+
+        emit Withdrawn(msg.sender, recipientLookup[depositId_], amountWithdrawn);
     }
 
     /**
-        @notice Convert agnostic value at current index to flat sOHM value
-        @dev Agnostic value earns rebases. Agnostic value is amount / rebase_index.
-             1e9 is because sOHM has 9 decimals.
-     */
-    function _fromAgnostic(uint256 amount_) internal view returns (uint256) {
-        return (amount_ * (IsOHM(sOHM).index())) / 1e9;
+        @notice Redeem available gOHM yield from a specific deposit
+        @param depositId_ Deposit ID to withdraw gOHM yield from
+    */
+    function _redeem(uint256 depositId_) internal returns (uint256 amountRedeemed) {
+        if (redeemDisabled) revert YieldDirector_RedeemsDisabled();
+        if (recipientLookup[depositId_] != msg.sender) revert YieldDirector_NotYourYield();
+
+        amountRedeemed = _redeemYield(depositId_);
+
+        if (depositInfo[depositId_].principalAmount == 0) {
+            _closeDeposit(depositId_, depositInfo[depositId_].depositor);
+
+            uint256[] storage receiptIds = recipientIds[msg.sender];
+            uint256 idsLength = receiptIds.length;
+
+            for (uint256 i = 0; i < idsLength; ++i) {
+                if (receiptIds[i] == depositId_) {
+                    // Remove id from recipient's ids array
+                    receiptIds[i] = receiptIds[idsLength - 1]; // Delete integer from array by swapping with last element and calling pop()
+                    receiptIds.pop();
+                    break;
+                }
+            }
+
+            delete recipientLookup[depositId_];
+        }
+
+        emit Redeemed(msg.sender, amountRedeemed);
+        emit Donated(depositInfo[depositId_].depositor, msg.sender, amountRedeemed);
     }
 
     /**
-        @notice Convert flat sOHM value to agnostic value at a given index value
-        @dev Agnostic value earns rebases. Agnostic value is amount / rebase_index.
-             1e9 is because sOHM has 9 decimals.
-     */
-    function _fromAgnosticAtIndex(uint256 amount_, uint256 index_) internal pure returns (uint256) {
-        return (amount_ * index_) / 1e9;
+        @notice Redeem all available gOHM yield from the vault
+    */
+    function _redeemAll() internal returns (uint256 amountRedeemed) {
+        if (redeemDisabled) revert YieldDirector_RedeemsDisabled();
+
+        amountRedeemed = 0;
+
+        uint256[] storage receiptIds = recipientIds[msg.sender];
+        uint256 idsLength = receiptIds.length;
+
+        // We iterate through the array back to front so that we can delete
+        // elements from the array without changing the locations of any
+        // entries that have not been checked yet
+        for (uint256 index = idsLength; index > 0; index--) {
```

Comment:
for better readability why not instead of:
for (uint256 index = idsLength; index > 0; index--)
we do:
for (uint256 index = idsLength-1; index >= 0; index--)

Then inside your loop you don't need to reference index - 1. You can just reference index.